### PR TITLE
Added checking for input nodes named key in UnionMemo.

### PR DESCRIPTION
**Pull request by dawidsula26**
Added check for input nodes named "key" in union-memo. Without this check, it is possible to create situation where one field in union-memo's output is used to store both key of elements and value from branch.

### CODE REVIEW

**Review by mproch — COMMENTED** [review]
Can we have a test for that?

**Comment by dawidsula26** [comment]
Error message, when two input nodes have identical names after sanitation, might need improving. I didn't have any better ideas for it.

File: engine/flink/components/base/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/UnionWithMemoTransformer.scala [hunk 1]
```diff
@@ -69,7 +71,20 @@ class UnionWithMemoTransformer(timestampAssigner: Option[TimestampWatermarkHandl
...
+    val validatedIdentical = ids.toList.groupBy(ContextTransformation.sanitizeBranchName).map {
+      case (_, values) if values.size >= 2 =>
+        val namesList = values.map("\"" + _ + "\"").mkString(", ")
+        ProcessCompilationError.CustomNodeError(s"Nodes $namesList have too similar names", None).invalidNel
+      case _ =>
+        Validated.validNel(Unit)
+    }.toList.sequence_
+
+    val validatedBranches = Validated.condNel(!ids.map(ContextTransformation.sanitizeBranchName).contains(KeyField), (),
+      ProcessCompilationError.CustomNodeError(s"""Input node can not be named \"$KeyField\"""", None))
```
Comment (mproch) [inline engine/flink/components/base/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/UnionWithMemoTransformer.scala:85]: With """ you don't need to escape "

File: engine/flink/components/base/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/UnionWithMemoTransformer.scala [hunk 2]
```diff
@@ -69,7 +71,20 @@ class UnionWithMemoTransformer(timestampAssigner: Option[TimestampWatermarkHandl
   def transformContextsDefinition(valueByBranchId: Map[String, LazyParameter[AnyRef]], variableName: String)
                                  (inputContexts: Map[String, ValidationContext])
                                  (implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, ValidationContext] = {
-    ContextTransformation.findUniqueParentContext(inputContexts).map { parent =>
+    val ids = valueByBranchId.keySet
+
+    val validatedIdentical = ids.toList.groupBy(ContextTransformation.sanitizeBranchName).map {
```
Comment (mproch) [inline engine/flink/components/base/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/UnionWithMemoTransformer.scala:76]: let's extract method "validateBranchNames"

File: engine/flink/components/base/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/UnionWithMemoTransformerSpec.scala [hunk 3]
```diff
@@ -92,6 +91,77 @@ class UnionWithMemoTransformerSpec extends FunSuite with FlinkSpec with Matchers
...
+    ))
+
+    val sourceFoo = BlockingQueueSource.create[OneRecord](_.timestamp, Duration.ofHours(1))
+    val sourceBar = BlockingQueueSource.create[OneRecord](_.timestamp, Duration.ofHours(1))
+
+    val collectingListener = ResultsCollectingListenerHolder.registerRun(identity)
+
+    val model = LocalModelData(ConfigFactory.empty(), new UnionWithMemoTransformerSpec.Creator(sourceFoo, sourceBar, collectingListener))
+    val validationResult = model.validator.validate(process).result
+    assert(validationResult.isInvalid)
```
Comment (mproch) [inline engine/flink/components/base/src/test/scala/pl/touk/nussknacker/engine/flink/util/transformer/UnionWithMemoTransformerSpec.scala:126]: maybe some assertion on error? e.g. using
https://www.scalatest.org/user_guide/using_matchers#matchingAPattern

File: components-api/src/main/scala/pl/touk/nussknacker/engine/api/context/ContextTransformation.scala [hunk 4]
```diff
@@ -56,6 +56,23 @@ object ContextTransformation {
...
+    } else {
+      List()
+    }
+  }
+
+  def checkSanitizedBranchNames(branchIds: Set[String])(implicit nodeId: NodeId): List[ProcessCompilationError] =
+    branchIds.toList.groupBy(sanitizeBranchName).flatMap{
+      case (_, values) if values.size >= 2 =>
+        val namesList = values.map("\"" + _ + "\"").mkString(", ")
+        List(ProcessCompilationError.CustomNodeError(s"Nodes $namesList have too similar names", None))
```
Comment (arkadius) [inline components-api/src/main/scala/pl/touk/nussknacker/engine/api/context/ContextTransformation.scala:71]: Maybe to be more precise "conflicting names"?

**Review by arkadius — APPROVED** [review]

**Pull request merged**